File: traceguard-extension/translate_script.py
import pyjson5
import json
import time
from deep_translator import GoogleTranslator

# Read the file
with open('src/lib/translations.ts', 'r', encoding='utf-8') as f:
    content = f.read()

# Extract the JSON object
prefix = "export const resources = "
json_str = content[content.find(prefix) + len(prefix):].strip()
if json_str.endswith(';'):
    json_str = json_str[:-1]

# Parse the JS object
data = pyjson5.decode(json_str)

# Get English keys from 'es'
english_keys = list(data['es']['translation'].keys())

# Initialize fr and de dictionaries
if 'fr' not in data:
    data['fr'] = {'translation': {}}
if 'de' not in data:
    data['de'] = {'translation': {}}

# Translate in chunks to avoid rate limits
def translate_list(texts, target_lang):
    translator = GoogleTranslator(source='en', target=target_lang)
    translated_dict = {}
    
    # Process in chunks
    chunk_size = 50
    for i in range(0, len(texts), chunk_size):
        chunk = texts[i:i+chunk_size]
        try:
            results = translator.translate_batch(chunk)
            for j, text in enumerate(chunk):
                translated_dict[text] = results[j]
            time.sleep(1) # Be nice to the API
        except Exception as e:
            print(f"Error translating to {target_lang} at chunk {i}: {e}")
            # fallback to individual translations
            for text in chunk:
                try:
                    res = GoogleTranslator(source='en', target=target_lang).translate(text)
                    translated_dict[text] = res
                    time.sleep(0.5)
                except Exception as ex:
                    print(f"Error on {text}: {ex}")
                    translated_dict[text] = text
    return translated_dict

print("Translating to French...")
fr_translations = translate_list(english_keys, 'fr')

print("Translating to German...")
de_translations = translate_list(english_keys, 'de')

data['fr']['translation'] = fr_translations
data['de']['translation'] = de_translations

# Convert back to TS file format
new_json_str = json.dumps(data, indent=2, ensure_ascii=False)
new_content = f"export const resources = {new_json_str};\n"

with open('src/lib/translations.ts', 'w', encoding='utf-8') as f:
    f.write(new_content)

print("Done!")

File: traceguard-extension/translate_script.mjs
import fs from 'fs';
import { translate } from '@vitalets/google-translate-api';

const filePath = 'src/lib/translations.ts';

async function run() {
  const fileContent = fs.readFileSync(filePath, 'utf8');
  const prefix = 'export const resources = ';
  let jsonStr = fileContent.substring(fileContent.indexOf(prefix) + prefix.length).trim();
  if (jsonStr.endsWith(';')) jsonStr = jsonStr.slice(0, -1);

  // We need to parse this string. Since it's not strict JSON, let's use eval or Function to get the object
  const data = new Function(`return ${jsonStr}`)();

  const englishKeys = Object.keys(data.es.translation);

  if (!data.fr) data.fr = { translation: {} };
  if (!data.de) data.de = { translation: {} };

  const delay = (ms) => new Promise(res => setTimeout(res, ms));

  async function translateBatch(keys, lang) {
    const results = {};
    // Chunking to avoid large requests
    const chunkSize = 20;
    for (let i = 0; i < keys.length; i += chunkSize) {
      const chunk = keys.slice(i, i + chunkSize);
      try {
        const textToTranslate = chunk.join('\n|||\n');
        const res = await translate(textToTranslate, { to: lang });
        const translations = res.text.split('\n|||\n');
        
        for (let j = 0; j < chunk.length; j++) {
          results[chunk[j]] = translations[j] ? translations[j].trim() : chunk[j];
        }
        await delay(500);
      } catch (err) {
        console.error(`Error chunk ${i} for ${lang}:`, err.message);
        // Fallback to individual
        for (let key of chunk) {
          try {
             const singleRes = await translate(key, { to: lang });
             results[key] = singleRes.text;
             await delay(200);
          } catch(e) {
             console.error(`Error on single ${key}:`, e.message);
             results[key] = key;
          }
        }
      }
    }
    return results;
  }

  console.log("Translating to French...");
  data.fr.translation = await translateBatch(englishKeys, 'fr');
  
  console.log("Translating to German...");
  data.de.translation = await translateBatch(englishKeys, 'de');

  const newContent = `${prefix}${JSON.stringify(data, null, 2)};\n`;
  fs.writeFileSync(filePath, newContent, 'utf8');
  console.log("Done!");
}

run().catch(console.error);
